### PR TITLE
[WIP]  State machine start line and header parsing

### PR DESCRIPTION
**Pull request by pakrym**

### CODE REVIEW

**Comment by pakrym** [comment]
This PR eliminates SIMD seeks and tries to parse everything in one pass through memory. 
During first tests showed ~2-3% improvement on big iron.

**Comment by cesarbs** [comment]
We discussed this some time ago and concluded the added complexity wasn't worth the effort. What changed?

**Comment by halter73** [comment]
If it's faster, functionally equivalent and it doesn't make the code harder to follow, I certainly won't complain 😄

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs [hunk 1]
```diff
@@ -77,6 +79,12 @@ public int Take()
                 return block.Array[index];
             }
 
+            return TakeSlow(wasLastBlock, block);
+        }
+
+        private int TakeSlow(bool wasLastBlock, MemoryPoolBlock block)
```
Comment (benaadams) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Infrastructure/MemoryPoolIterator.cs]: Move `wasLastBlock` into method, isn't used in fast path

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 2]
```diff
@@ -928,129 +942,271 @@ public RequestLineStatus TakeStartLine(SocketInput input)
...
-                        RejectRequest(RequestRejectionReason.InvalidRequestLine,
-                            Log.IsEnabled(LogLevel.Information) ? start.GetAsciiStringEscaped(end, MaxInvalidRequestLineChars) : string.Empty);
-                    }
+                MemoryPoolIterator verbEnd = start;
+                MemoryPoolIterator pathStart = start;
+                MemoryPoolIterator pathEnd = start;
+                MemoryPoolIterator queryStart = start;
+                MemoryPoolIterator versionStart = start;
+                MemoryPoolIterator queryEnd = start;
+                MemoryPoolIterator versionEnd = start;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: `var`

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 3]
```diff
@@ -928,129 +942,271 @@ public RequestLineStatus TakeStartLine(SocketInput input)
...
+                                }
+                                else if (ch != -1)
+                                {
+                                    queryEnd = scan;
+                                    queryRead = true;
+                                }
+                            } while (!scan.IsEnd && limit > 0);
+                            break;
+                        case StartLineParsingState.VersionFast:
+                        {
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Why does only this `case` have a block? Why is the break outside the block?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 4]
```diff
@@ -928,129 +942,271 @@ public RequestLineStatus TakeStartLine(SocketInput input)
...
+                            int bytes;
 
-                    if (httpVersion == string.Empty)
+                            line.Seek(ref _vectorLFs, out bytes, ServerOptions.Limits.MaxRequestLineSize);
+                            line.Skip(1);
+                            RejectRequest(RequestRejectionReason.InvalidRequestLine,
+                                start.GetAsciiStringEscaped(line, MaxInvalidRequestLineChars));
+                            break;
+                        default:
+                            throw new ArgumentOutOfRangeException();
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: Is this the recommended exception type for this situation?

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 5]
```diff
@@ -1174,159 +1330,214 @@ private bool RequestUrlStartsWithPathBase(string requestUrl, out bool caseMatche
...
+                                        consumed = scan;
+                                        return true;
+                                    }
+
+                                    // Headers don't end in CRLF line.
+                                    RejectRequest(RequestRejectionReason.HeadersCorruptedInvalidHeaderSequence);
+                                }
+                                break;
+                            case HeaderParsingState.Name:
+                                bool nameRead = false;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: `var`

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs [hunk 6]
```diff
@@ -1174,159 +1330,214 @@ private bool RequestUrlStartsWithPathBase(string requestUrl, out bool caseMatche
...
+                                    limit--;
+                                    if (ch == '\r')
+                                    {
+                                        state = HeaderParsingState.Newline;
+                                        break;
+                                    }
+                                    else if (ch != -1)
+                                    {
+                                        //valueRead = true;
+                                        if (ch != ' ' && ch != '\t')
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/Frame.cs]: `&&` these in the `else if` above

**Comment by mikeharder** [comment]
Can we write a micro benchmark on sample input using Benchmark.NET or similar?

**Pull request closed without merging**